File: rin.Framework/Graphics/DeviceBuffer.cs
using System.Runtime.InteropServices;
using rin.Framework.Core;
using rin.Framework.Graphics.FrameGraph;
using TerraFX.Interop.Vulkan;
using static TerraFX.Interop.Vulkan.Vulkan;
namespace rin.Framework.Graphics;

/// <summary>
///     GPU Buffer
/// </summary>
public class DeviceBuffer : DeviceMemory, IDeviceBuffer
{
    private ulong? _address;
    
    public string Name { get; private set; }
    /// <summary>
    ///     GPU Buffer
    /// </summary>
    public DeviceBuffer(VkBuffer inBuffer, ulong inSize, Allocator allocator, IntPtr allocation, string name) : base(allocator,
        allocation)
    {
        NativeBuffer = inBuffer;
        Size = inSize;
        Name = name;
    }


    public static implicit operator VkBuffer(DeviceBuffer from)
    {
        return from.NativeBuffer;
    }

    protected override void OnDispose(bool isManual)
    {
        Allocator.FreeBuffer(this);
    }

    public ulong Offset => 0;
    public ulong Size { get; }

    public VkBuffer NativeBuffer { get; }

    public ulong GetAddress()
    {
        if (_address.HasValue) return _address.Value;
        
        var info = new VkBufferDeviceAddressInfo()
        {
            sType = VkStructureType.VK_STRUCTURE_TYPE_BUFFER_DEVICE_ADDRESS_INFO,
            buffer = NativeBuffer
        };
        unsafe
        {
            _address = vkGetBufferDeviceAddress(SGraphicsModule.Get().GetDevice(), &info);
        }

        return _address.Value;
    }

    public IDeviceBufferView GetView(ulong offset, ulong size)
    {
        return new DeviceBufferView(this, offset, size);
    }

    public unsafe void Write(void* src, ulong size, ulong offset = 0)
    {
        NativeMethods.CopyToBuffer(Allocator, Allocation.ToPointer(), src,size,offset);
    }
}

File: lib/rin/rhi/DeviceBuffer.cpp
#include "rin/rhi/DeviceBuffer.h"
#include "vk_mem_alloc.h"
#include "rin/core/macros.h"
#include "rin/rhi/DeviceBufferView.h"
#include "rin/rhi/GraphicsModule.h"

namespace rin::rhi
{
    void DeviceBuffer::OnDispose()
    {
        vmaDestroyBuffer(_allocator,_buffer,_allocation);
    }

    DeviceBuffer::DeviceBuffer(const VmaAllocator& allocator, const VmaAllocation& allocation, const vk::Buffer& buffer,
        const uint64_t& size)
    {
        _allocator = allocator;
        _allocation = allocation;
        _buffer = buffer;
        _size = size;
    }

    uint64_t DeviceBuffer::GetOffset()
    {
        return 0;
    }

    uint64_t DeviceBuffer::GetSize()
    {
        return _size;
    }

    vk::Buffer DeviceBuffer::GetBuffer()
    {
        return _buffer;
    }

    uint64_t DeviceBuffer::GetAddress()
    {
        NOT_IMPLEMENTED
    }

    Shared<IDeviceBuffer> DeviceBuffer::GetView(uint64_t offset, uint64_t size)
    {
        return shared<DeviceBufferView>(shared_from_this(), offset, size);
    }

    void DeviceBuffer::Write(void* src, uint64_t size, uint64_t offset)
    {
        vmaCopyMemoryToAllocation(_allocator,src,_allocation,offset,size);
    }
}
